### PR TITLE
Add timestamp and L1 block info to sequencer blocks

### LINKED ISSUE #536: Add L1 block number and timestamp to `next_block`

### PR DESCRIPTION
**Pull request by jbearer**
Note: I think a better design for this would be to have some shared memory (ie a `Mutex` or channel) where we store the latest known L1 block, and have an async task to update this state each time it receives a new block from the L1 provider. Then we could move L1 queries, and the complexities that occur when they are slow or fail, from the critical path of proposal; proposal would only need to read from the `Mutex`. Unfortunately, since this design is stateful, it can't really be implemented with the current HotShot traits. Consider this change a short term workaround.

The timestamp is needed because we work in terms of block commitments,
not leaf commitments, so clients cannot authenticate the timestamp
from the leaf. Perhaps we should change this.
    
I had to do a pretty ugly hack to set up an L1 provider and get the
latest L1 block when proposing a new block. There are some (I think
simple) changes to HotShot APIs that would make this much cleaner.
    
Finally, ntoe that this is not secure. The proposer has full
discretion to choose this metadata, and safety invariants (like
monotonicity) as well as performance requirements (like using a
sufficiently recent L1 block) are not currently enforced. The plan
is to eventually hook validation code into HotShot replicas via
the metastate API.
    
Closes #536

### CODE REVIEW

**Comment by jbearer** [comment]
Some weird issues with this change, debugging

**Comment by jbearer** [comment]
Fixed (I think)

File: sequencer/src/state.rs [hunk 1]
```diff
@@ -20,11 +26,79 @@ impl HotShotState for State {
 
     type Time = ViewNumber;
 
-    fn next_block(prev_state: Option<Self>) -> Self::BlockType {
-        match prev_state {
-            Some(_state) => unimplemented!("Using sequencing consensus, no previous state"),
-            None => Block::new(),
-        }
+    fn next_block(_prev_state: Option<Self>) -> Self::BlockType {
```
Comment (nomaxg) [inline sequencer/src/state.rs:29]: Ooof yeah these all seem like necessary improvements to the API

File: sequencer/src/state.rs [hunk 2]
```diff
@@ -20,11 +26,79 @@ impl HotShotState for State {
...
+                        block.timestamp
+                    );
+                    break L1BlockInfo {
+                        number: number.as_u64(),
+                        timestamp: block.timestamp,
+                    };
+                }
+            })
+        } else {
+            // For unit testing, we may disable the L1 provider and use mock L1 blocks instead.
```
Comment (sveitser) [inline sequencer/src/state.rs:97]: Should we hide this behavior behind the `testing` feature flag?

File: sequencer/src/api.rs [hunk 3]
```diff
@@ -295,17 +297,17 @@ impl<N: network::Type> Borrow<HotShotSequencingConsensusApi<SeqTypes, Node<N>>>
 
 #[derive(Clone, Debug, Serialize, Deserialize)]
 pub struct NamespaceProofQueryData {
-    proof: NamespaceProofType,
-    nmt_root: NMTRoot,
+    pub proof: NamespaceProofType,
+    pub header: Header,
```
Comment (sveitser) [inline sequencer/src/api.rs:301]: What's the reason for making these public if we also have getters?

**Review by sveitser — APPROVED** [review]
LGTM, some minor questions.

File: sequencer/src/api.rs (hunk 3 above)
Comment (jbearer) [reply sequencer/src/api.rs:301]: So that you can move out of the fields

File: sequencer/src/state.rs (hunk 2 above)
Comment (jbearer) [reply sequencer/src/state.rs:97]: Yes, good idea

**Pull request merged**